engine/table/src/main/java/io/deephaven/engine/table/impl/AbstractFilterExecution.java

```diff
@@ -232,77 +205,425 @@ public LogOutput append(LogOutput output) {
+    }
+
+    /**
+     * Simple extensions to hold either stateless or stateful filter.
+     */
+    private static class FilterCollection extends ArrayList<WhereFilter> {
+    }
+    private static class StatelessFilterCollection extends FilterCollection {
+    }
+    private static class StatefulFilterCollection extends FilterCollection {
+    }
+
+    /**
+     * Transform an array of {@link WhereFilter filters} into collections of stateful and stateless filters.
      *
-     * @param onComplete the routine to call after the filter has been completely executed.
-     * @param onError the routine to call if the filter experiences an exception.
+     * @param filters the filters to collect
+     * @return a list of filter collections
      */
-    public void scheduleCompletion(
-            @NotNull final AbstractFilterExecution.FilterComplete onComplete,
-            @NotNull final Consumer<Exception> onError) {
+    static List<FilterCollection> collectFilters(WhereFilter[] filters) {
+        final List<FilterCollection> filterCollections = new ArrayList<>();
+        boolean collectionStateless = true;
+        for (final WhereFilter filter : filters) {
+            final boolean filterStateless = filter.permitParallelization(); // determines if a filter is stateless
+            if (filterCollections.isEmpty() || collectionStateless != filterStateless) {
+                filterCollections.add(filterStateless
+                        ? new StatelessFilterCollection()
+                        : new StatefulFilterCollection());
+                collectionStateless = filterStateless;
+            }
+            filterCollections.get(filterCollections.size() - 1).add(filter);
+        }
+        return filterCollections;
+    }
 
-        // Start with the input row sets and narrow with each filter.
-        final MutableObject<WritableRowSet> localAddInput = new MutableObject<>(
-                addedInput != null && addedInput.isNonempty()
-                        ? addedInput.copy()
-                        : null);
-        final MutableObject<WritableRowSet> localModInput = new MutableObject<>(
-                runModifiedFilters && modifiedInput != null && modifiedInput.isNonempty()
-                        ? modifiedInput.copy()
-                        : null);
-        if (localAddInput.getValue() == null && localModInput.getValue() == null) {
-            onComplete.accept(RowSetFactory.empty(), RowSetFactory.empty());
+    /**
+     * Execute the final filter (as opposed to the pushdown pre-filters) and pass the result to the consumer.
+     */
+    private void executeFinalFilter(
+            final WhereFilter filter,
+            final RowSet input,
+            final Consumer<WritableRowSet> resultConsumer,
+            final Consumer<Exception> exceptionConsumer) {
+        if (input.isEmpty()) {
+            // If there are no rows left to filter, skip this filter.
+            resultConsumer.accept(RowSetFactory.empty());
```

Comment:
We didn't hit this block in coverage.

Reply:
This check is unnecessary and has been removed. `filterStatelessCollection()` and `filterStatefulCollection()` both check before each iteration for an empty input state. Also, if pushdown filtering eliminates all `maybe` rows we terminate early (in `executeStatelessFilter()`).